Assorted updates for C++17

nuoHep opened the pull request:
The main thing here is replacing of our custom `string_view` with the `std` one. I left the `util` alias in place to minimize the diff.
Next step if no one minds the churn could be to remove that and replace all `util::` and `fmt::` `string_view`s with `std::`.

Also some other stuff I remembered could be done with the update to c++17.

nuoHep committed:
Drop range::invoke()

Use c++17 std::invoke instead.

nuoHep committed:
Drop meta::static_const_t

Use c++17 inline constexpr variables.

nuoHep committed:
Util: static_map - drop custom array implementation

Switch to std::array. Possible now as it has constexpr non-const
operator[] in c++17.

nuoHep committed:
Util: switch to std::string_view

Drop our internal implementation of string_view in favor of the std one.

Leave the util:: alias for now to minimize the diff.

nuoHep committed:
Drop ad-hoc void_t helpers

Use c++17 std::void_t instead

scamille commented:
Great to see that we can have that upgrade without actually having the additional c++20 string-view functions in the standard.

nuoHep committed:
Expressions: update for std::string_view transition

Simply return EOF token here without going through yield_token() as all
of the string manipulations there are meaningless for EOF.

The actual bug here is that for EOF current_len still contains the
length of the previous token, so remove_prefix() in yield_token() gets
called with a non-zero n while the input string is empty. This is UB for
std::string_view and all of the stls assert for it (in some way or
another) while our implementation of string_view simply clamps the
passed in value.

scamille commented on engine/util/static_map.hpp:
```diff
@@ -119,24 +108,24 @@ class static_set_base {
 
   constexpr iterator end() const { return cend(); }
-  constexpr iterator cend() const { return data_.end(); }
+  constexpr iterator cend() const { return data_.data() + N; }
```
Is there a reason to use data() and data()+N here instead of directly using the corresponding std::array functions?

nuoHep replied:
Yeah, different types.

`static_xxx::iterator` is a pointer. `std::array<>::iterator` is implementation defined, just has to satisfy some named requirements.
So while it [`std::array<>::iterator`] is *typically* a pointer, at least in release builds, in msvc debug builds it's a checked iterator struct.

I actually had it call the corresponding `std::array` functions initially, but had to change it (https://github.com/simulationcraft/simc/compare/212a90813c8a97d550e6799fbda4c08e073c3a44..66029fa2b3ec5140a0219d2bfb35a60efda44b0a) because of that.
